schemas/kargo/project_types.cue

```diff
@@ -0,0 +1,152 @@
+package kargo
+
+import (
+	holos "example.com/platform/schemas/holos/v1alpha5"
+	stage "kargo.akuity.io/stage/v1alpha1"
+)
+
+// #Projects defines the structure of a kargo project, useful for kargo
+// related components to look up data given a ProjectName.
+#Projects: [NAME=string]: #Project & {name: NAME}
+
+// #Project defines the structure of a project configured for progressive
+// rollouts with Kargo.  The structure composes one Holos #Project with one
+// Kargo #Project.
+//
+// We define an abstraction over both kinds of projects, holos and kargo,
+// because the two are closely related but not the same.  We want to keep the
+// concerns separate with loose coupling.
+//
+// The Holos Project is used to associate multiple related components together
+// and isn't concerned with Kargo.
+//
+// The Kargo Project is used to define a progressive rollout promotion pipeline
+// across multiple stages.  For example, Kargo handles the business use case:
+// automatically promote from dev to test to uat, then submit a pull request to
+// promote to production.  On merge, roll out to production across regions from
+// east to west, verifying each deployment is healthy before proceeding to the
+// next.
+//
+// See the #ProjectBuilder for how concrete values of this schema are built from
+// a collection of components and stages.
+#Project: {
+	name:  #Name
+	#Name: string
+
+	stages: holos.#Stages
+
+	// promotions maps the promotable component names to pipeline stages.
+	promotions: [holos.#Component.#Name]: {
+		requestedFreight: stage.#StageSpec.requestedFreight
+	}
+
+	// Automatically promote non-prod stages.
+	promotionPolicies: [for STAGE in stages if STAGE.tier == "nonprod" {
+		stage:                "\(STAGE.name)-\(name)"
+		autoPromotionEnabled: true
+	}]
+
+	// Compose the holos Project into the kargo project.
+	HolosProject: holos.#Project
+}
+
+// ProjectBuilder expands components out across the provided stages and
+// configures a Kargo Project to manage the promotion process across stages for
+// the components in the project.
+#ProjectBuilder: {
```

Comment:
Here's the Kargo ProjectBuilder, which is the guts of this.  It also assembles a holos Project to go along with the kargo Project.